keps/sig-scheduling/3280-guarantee-pdb-when-preemption-happens/README.md

```diff
@@ -0,0 +1,981 @@
+has a priority ClassName of `system-cluster-critical` or `system-node-critical`, it may 
+still potentially violate the victim's PDB. New policies may be added in the future based 
+on the need, but it will not be discussed in this article at this time.
+
+```go
+// DisruptionPolicy specifies how pods with this priority class 
+// may be disrupted when being preempted.
+type DisruptionPolicy string
+
+const (
+	// SystemCriticalOnly indicates when being preempted by a higher priority pod, 
+	// strict adherence to the PodDisruptionBudget of the pod with this priority class is required.
+	// But it the preempter with these priority classes of `system-cluster-critical` or `system-node-critical` 
+	// will still be able to violate this pod's PDB during preemption.
+	SystemCriticalOnly DisruptionPolicy = "SystemCriticalOnly"
+	
+	// BestEffort indicates when being preempted by a higher priority pod, 
+	// an effort is made to adhere to the PodDisruptionBudget of the pod with this priority class
+	// as much as possible.
+	BestEffort DisruptionPolicy = "BestEffort"
+)
+```
+
+The following is an example of a `PriorityClass` (PDB) where the user sets the `disruptionPolicy` 
+to `PodDisruptionBudgetStrict` indicating that the PDB for the zookeeper must not be violated during the 
+scheduler preemption process.
+
+```yaml
+apiVersion: scheduling.k8s.io/v1
+kind: PriorityClass
+metadata:
+  name: low-priority
+value: 100
+globalDefault: false
+description: "This priority class should be used for XYZ service pods only."
+disruptionPolicy: SystemCriticalOnly
+```
+
+The scheduler plugin `defaultpreemption` needs to check the policy set in the "preemptionPolicy" field 
+when selecting victims.
+- if it's set to `PodDisruptionBudgetBestEffort`, the implementation will remain consistent with the existing behavior,
+  meaning that the scheduler will try to select victims whose PDBs are not violated by preemption, but 
+  if no such victims are found, preemption will still happen and lower priority pods will be removed 
+  despite their PDBs being violated.
+- if it's set to `PodDisruptionBudgetStrict`, there are two options:
+  - If the preempter's priority ClassName is `system-cluster-critical` or `system-node-critical` the implementation 
+    will be the same as `PodDisruptionBudgetBestEffort`.
+  - If not, the scheduler will check if the victim' PDBs will be violated when selecting victims
+    - if violate the victims' PDBs, this victim will not be selected as candidates.
+    - if not violate the victims' PDBs, scheduler will preempt this pod via the `/evictions` endpoint. If it responds `200 OK`, 
+      it means the eviction is allowed, and the victim is deleted, similar to sending a DELETE request to the Pod URL. If it responses
+      `429 Too Many Requests` or `500 Internal Server Error`, the scheduler will output an error log and terminate the preemption process 
```

Comment:
Shouldn't we retry here on other preemption candidates here?
@alculquicondor

Reply:
Yes. we can retry on other candidates. Updated by the https://github.com/kubernetes/enhancements/pull/3755#discussion_r1094868040